src/e_ibmca.c

```diff
@@ -2326,6 +2395,407 @@ static int ibmca_cipher_cleanup(EVP_CIPHER_CTX * ctx)
+}
+
+static int ibmca_aes_gcm(ICA_AES_GCM_CTX *ctx, const unsigned char *in,
+                         unsigned char *out, size_t len, int enc, int keylen)
+{
+	uint64_t mlen = ctx->ptlen;
+
+	mlen += len;
+	if (mlen > ((1ULL << 36) - 32) || (sizeof(len) == 8 && mlen < len))
+		return -1;
+
+	ctx->ptlen = mlen;
+
+	/* ctx->taglen is not set at this time... and not needed. The function
+	 * only checks, if it's a valid gcm tag length. so we chose use 16. */
+	return p_ica_aes_gcm_intermediate((unsigned char *)in, len, out,
+					  ctx->iv, NULL, 0, ctx->tag, 16,
+					  ctx->key, keylen, ctx->subkey, enc);
+}
+
+static int ibmca_aes_gcm_init_key(EVP_CIPHER_CTX *ctx,
+                                  const unsigned char *key,
+                                  const unsigned char *iv, int enc)
+{
+	ICA_AES_GCM_CTX *gctx =
+#ifdef OLDER_OPENSSL
+	    ctx->cipher_data;
+#else
+	    (ICA_AES_GCM_CTX *)EVP_CIPHER_CTX_get_cipher_data(ctx);
+#endif
+	const int gkeylen = EVP_CIPHER_CTX_key_length(ctx);
+	unsigned char *goiv =
+#ifdef OLDER_OPENSSL
+	    ctx->oiv;
+#else
+	    (unsigned char *)EVP_CIPHER_CTX_original_iv(ctx);
+#endif
+	unsigned char *giv =
+#ifdef OLDER_OPENSSL
+		    ctx->iv;
+#else
+		    EVP_CIPHER_CTX_iv_noconst(ctx);
+#endif
+	ica_aes_vector_t zeroiv;
+
+	memset(zeroiv, 0, sizeof(zeroiv));
+
+	if (!iv && !key)
+	    return 1;
+
+	if (key) {
+		memcpy(gctx->key, key, gkeylen);
+		memset(gctx->tag, 0, sizeof(gctx->tag));
+		memset(gctx->subkey, 0, sizeof(gctx->subkey));
+		gctx->aadlen = 0;
+		gctx->ptlen = 0;
+
+		if (iv == NULL && gctx->iv_set)
+			iv = gctx->iv;
+
+		if (iv) {
+			p_ica_aes_gcm_initialize(iv, gctx->ivlen,
+						 (unsigned char *)key,
+						 gkeylen, zeroiv, goiv,
+						 gctx->subkey, enc);
+			memcpy(giv, goiv, sizeof(zeroiv));
+			gctx->iv_set = 1;
+		}
+		gctx->key_set = 1;
+	} else {
+		if (gctx->key_set) {
+			p_ica_aes_gcm_initialize(iv, gctx->ivlen,
+						 (unsigned char *)key,
+						 gkeylen, zeroiv, goiv,
+						 gctx->subkey, enc);
+			memcpy(giv, goiv, sizeof(zeroiv));
+		} else {
+			memcpy(gctx->iv, iv, gctx->ivlen);
+		}
+		gctx->iv_set = 1;
+		gctx->iv_gen = 0;
+	}
+	return 1;
+}
+
+static int ibmca_aes_gcm_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg,
+			      void *ptr)
+{
+	ICA_AES_GCM_CTX *gctx =
+#ifdef OLDER_OPENSSL
+	    ctx->cipher_data;
+#else
+	    (ICA_AES_GCM_CTX *)EVP_CIPHER_CTX_get_cipher_data(ctx);
+#endif
+	unsigned char *iv_noconst =
+#ifdef OLDER_OPENSSL
+		    ctx->iv;
+#else
+		    EVP_CIPHER_CTX_iv_noconst(ctx);
+#endif
+	int enc =
+#ifdef OLDER_OPENSSL
+		    ctx->encrypt;
+#else
+		    EVP_CIPHER_CTX_encrypting(ctx);
+#endif
+
+	switch (type) {
+	case EVP_CTRL_INIT:
+		gctx->key_set = 0;
+		gctx->iv_set = 0;
+		gctx->ivlen = EVP_CIPHER_CTX_iv_length(ctx);
+		gctx->iv = iv_noconst;
+		gctx->taglen = -1;
+		gctx->iv_gen = 0;
+		gctx->tls_aadlen = -1;
+		return 1;
+
+#ifdef OLDER_OPENSSL
+	case EVP_CTRL_GCM_SET_IVLEN:
+#else
+	case EVP_CTRL_AEAD_SET_IVLEN:
+#endif
```

Comment:
Why not do something like
```
		return 1;

	case EVP_CTRL_GCM_SET_IVLEN:
	case EVP_CTRL_AEAD_SET_IVLEN:
		if (arg <= 0)
```
or the compiler will fail?  With that, the block will be executed in one or other case.